syntax = "proto3";

package git;

message Repo {
    string name = 1;
    string user = 2;
}

service ManipulateRepo{
    rpc CreateRepo(Repo) returns (Repo);
}

message Branch{
    Repo repo =1;
    string name =2;
}

message Branches{
    Repo repo = 1;
    repeated string branches=2;
}

service ManipulateBranches{
    rpc ListBranches(Repo) returns (Branches);
}

message Commit {
    Repo repo = 1;
    string hash = 2;
    string hash_abbv = 3;
    string tree_hash = 4;
    string tree_hash_abbv = 5;
    string parent_hash = 6;
    string parent_hash_abbv = 7;
    string commit_message = 8;
    int64 date = 9;
}

message Commits {
    repeated Commit commits = 1;
}

service ManipulateCommits {
    rpc GetHeadCommit (Repo) returns (Commit);
    rpc FindAllCommits (Repo) returns (Commits);
}

message File {
    Branch branch = 1;
    string name = 2;
    string type = 3; // Directory, File,
    string directory =4;
}

message Files {
    repeated File files =1;
}

service ManipulateFiles{
    rpc GetHeadCommitFiles (Branch) returns (Files);
    rpc GetMasterHeadCommitFiles(Repo) returns (Files);
}